AU2/plugins/custom_plugins/PolicePlugin.py

```diff
@@ -0,0 +1,228 @@
+from AU2.plugins.CorePlugin import registered_plugin
+from AU2.plugins.constants import COLLEGES, WATER_STATUSES, WEBPAGE_WRITE_LOCATION
+from AU2.plugins.util import random_data
+from AU2.plugins.util.DeathManager import DeathManager
+from AU2.plugins.util.PoliceRankManager import DEFAULT_RANKS, PoliceRankManager
+
+POLICE_TABLE_TEMPLATE = """
+<p xmlns="">
+      Here is a list of members of our loyal police, charged with protecting Cambridge from murderers of innocents and people that annoy the all powerful Umpire:
+   </p>
+<table xmlns="" class="playerlist">
+  <tr><th>Rank</th><th>Pseudonym</th><th>Real Name</th><th>Email Address</th><th>College</th><th>Notes</th></tr>
+  {ROWS}
+</table>
+"""
+
+POLICE_TABLE_ROW_TEMPLATE = """
+<tr><td>{RANK}</td><td>{PSEUDONYM}</td><td>{NAME}</td><td>{EMAIL}</td><td>{COLLEGE}</td><td>{NOTES}</td></tr>
+"""
+
+DEAD_POLICE_TABLE_TEMPLATE = """
+<p xmlns="">
+      Killed in the line of duty:
+   </p>
+<table xmlns="" class="playerlist">
+  <tr><th>Rank</th><th>Pseudonym</th><th>Real Name</th><th>Email Address</th><th>College</th><th>Notes</th></tr>
+  {ROWS}
+</table>
+"""
+
+DEAD_POLICE_TABLE_ROW_TEMPLATE = """
+<tr><td>{RANK}</td><td>{PSEUDONYM}</td><td>{NAME}</td><td>{EMAIL}</td><td>{COLLEGE}</td><td>{NOTES}</td></tr>
+"""
+
+DEAD_CORRUPT_POLICE_TABLE_TEMPLATE = """
+<p xmlns="">
+      And these are the names of those who turned to the dark side, and paid the price:
+   </p>
+<table xmlns="" class="playerlist">
+  <tr><th>Rank</th><th>Pseudonym</th><th>Real Name</th><th>Email Address</th><th>College</th><th>Notes</th></tr>
+  {ROWS}
+</table>
+"""
+
+DEAD_CORRUPT_POLICE_TABLE_ROW_TEMPLATE = """
+<tr><td>{RANK}</td><td>{PSEUDONYM}</td><td>{NAME}</td><td>{EMAIL}</td><td>{COLLEGE}</td><td>{NOTES}</td></tr>
+"""
+
+NO_DEAD_POLICE = """<p xmlns="">No police have been killed... yet.</p>"""
+
+POLICE_PAGE_TEMPLATE: str
+with open(os.path.join(ROOT_DIR, "plugins", "custom_plugins", "html_templates", "police.html"), "r") as F:
+    POLICE_PAGE_TEMPLATE = F.read()
+
+@registered_plugin
+class PolicePlugin(AbstractPlugin):
+    def __init__(self):
+        super().__init__("PolicePlugin")
+
+        self.html_ids = {
+            "Rank": self.identifier + "_rank",
+            "Rank Name": self.identifier + "_name"
+        }
+
+        self.config_exports = [
+            ConfigExport(
+                "PolicePlugin_set_ranks",
+                "Police -> Set Ranks",
+                self.ask_set_ranks,
+                self.answer_set_ranks
+            )
+        ]
+
+        self.ranks = GENERIC_STATE_DATABASE.arb_state.setdefault(self.identifier, DEFAULT_RANKS)
+
+    def ask_set_ranks(self):  # TODO Make Ctrl-C work - not sure why it isn't
+        return [
+            Label("Rename different ranks of police to fit the game theme"),
+            Label("Ranks affect position on the police page, and can be used to reward/punish/just have a bit of fun"),
+            InputWithDropDown(
```

Comment:
How about a [LargeTextEntry](https://github.com/jsyiek/AU2/blob/master/AU2/html_components/LargeTextEntry.py)? You could have each rank delimited by a newline. This would make bulk modification easy (as opposed to modding one at a time), and would allow you to have endless ranks.

This would mean you'd use a list and not a dictionary to deal with ranks. Your default ranks could then by `[Constable, Chief of Police, Umpire]` with the latter two hard-coded never to be reached by players. The rank up would then be a relative offset (so ask them how many ranks up or down to offset a police).

Reply:
Would ranking down (to below default) work then, and if so, how would the LargeTextEntry recognise the default?

Reply:
One possible solution is to make the "default default" the lowest rank list. And then if users want to override that, allow them to append `=default` to the end of the ranks. You can deal with it when parsing the output of `LargeTextEntry`.